backends/cloud/box/boxlistdirectorybyidrequest.cpp

```diff
@@ -0,0 +1,169 @@
+#include "backends/cloud/storage.h"
+#include "backends/networking/curl/connectionmanager.h"
+#include "backends/networking/curl/curljsonrequest.h"
+#include "backends/networking/curl/networkreadstream.h"
+#include "common/json.h"
+
+namespace Cloud {
+namespace Box {
+
+#define BOX_LIST_DIRECTORY_LIMIT 1000
+
+BoxListDirectoryByIdRequest::BoxListDirectoryByIdRequest(BoxStorage *storage, Common::String id, Storage::ListDirectoryCallback cb, Networking::ErrorCallback ecb):
+	Networking::Request(nullptr, ecb), _requestedId(id), _storage(storage), _listDirectoryCallback(cb),
+	_workingRequest(nullptr), _ignoreCallback(false) {
+	start();
+}
+
+BoxListDirectoryByIdRequest::~BoxListDirectoryByIdRequest() {
+	_ignoreCallback = true;
+	if (_workingRequest) _workingRequest->finish();
+	delete _listDirectoryCallback;
+}
+
+void BoxListDirectoryByIdRequest::start() {
+	_ignoreCallback = true;
+	if (_workingRequest) _workingRequest->finish();
+	_files.clear();
+	_ignoreCallback = false;
+
+	makeRequest(0);
+}
+
+void BoxListDirectoryByIdRequest::makeRequest(uint32 offset) {
+	Common::String url = Common::String::format(
+		"https://api.box.com/2.0/folders/%s/items?offset=%u&limit=%u&fields=%s",
+		_requestedId.c_str(),
+		offset,
+		BOX_LIST_DIRECTORY_LIMIT,
+		"id,type,name,size,modified_at"
+	);
+
+	Networking::JsonCallback callback = new Common::Callback<BoxListDirectoryByIdRequest, Networking::JsonResponse>(this, &BoxListDirectoryByIdRequest::responseCallback);
+	Networking::ErrorCallback failureCallback = new Common::Callback<BoxListDirectoryByIdRequest, Networking::ErrorResponse>(this, &BoxListDirectoryByIdRequest::errorCallback);
+	Networking::CurlJsonRequest *request = new BoxTokenRefresher(_storage, callback, failureCallback, url.c_str());
+	request->addHeader("Authorization: Bearer " + _storage->accessToken());
+	_workingRequest = ConnMan.addRequest(request);
+}
+
+void BoxListDirectoryByIdRequest::responseCallback(Networking::JsonResponse response) {
+	_workingRequest = nullptr;
+	if (_ignoreCallback) return;
+	if (response.request) _date = response.request->date();
+
+	Networking::ErrorResponse error(this);
+	Networking::CurlJsonRequest *rq = (Networking::CurlJsonRequest *)response.request;
+	if (rq && rq->getNetworkReadStream())
+		error.httpResponseCode = rq->getNetworkReadStream()->httpResponseCode();
+
+	Common::JSONValue *json = response.value;
+	if (json) {
+		Common::JSONObject responseObject = json->asObject();
+
+		//debug("%s", json->stringify(true).c_str());
+
+		//TODO: check that error is returned the right way
+		/*
+		if (responseObject.contains("error") || responseObject.contains("error_summary")) {
+		    warning("Box returned error: %s", responseObject.getVal("error_summary")->asString().c_str());
+		    error.failed = true;
+		    error.response = json->stringify();
+		    finishError(error);
+		    delete json;
+		    return;
+		}
+		*/
+
+		//TODO: check that ALL keys exist AND HAVE RIGHT TYPE to avoid segfaults
```

Comment:
Right, it is a must for implementing. Could also help with testing if the providers change their API.